File: data/Algorithms/Binary Search Tree.c
#include <stdio.h>
#include <conio.h>
#include <stdlib.h>
#include <malloc.h>

struct node
{
    int info;
    node *left, *right;
} * root, *ptr;

void inorder(node *tree)
{
    if (tree != NULL)
    {
        inorder(tree->left);
        printf("%d  ", tree->info);
        inorder(tree->right);
    }
}

void preorder(node *tree)
{
    if (tree != NULL)
    {
        printf("%d  ", tree->info);
        preorder(tree->left);
        preorder(tree->right);
    }
}

void postorder(node *tree)
{
    if (tree != NULL)
    {
        postorder(tree->left);
        postorder(tree->right);
        printf("%d  ", tree->info);
    }
}

node *insertNode(node *tree, int item)
{
    if (tree == NULL)
    {
        tree = (node *)malloc(sizeof(node));
        tree->left = tree->right = NULL;
        tree->info = item;
    }
    else
    {
        if (item < tree->info)
            tree->left = insertNode(tree->left, item);
        else if (item > tree->info)
            tree->right = insertNode(tree->right, item);
        else
            printf(" Duplicate Element !! Not Allowed !!!");
    }
    return tree;
}

node *minValueNode(node *subtree)
{
    node *current = subtree;
    while (current->left != NULL)
        current = current->left;
    return current;
}

node *deleteNode(node *root, int item)
{
    if (root == NULL)
        return root;

    if (item < root->info)
        root->left = deleteNode(root->left, item);

    else if (item > root->info)
        root->right = deleteNode(root->right, item);

    else
    {
        if (root->left == NULL)
        {
            node *temp = root->right;
            free(root);
            return temp;
        }
        else if (root->right == NULL)
        {
            node *temp = root->left;
            free(root);
            return temp;
        }

        node *temp = minValueNode(root->right);

        root->info = temp->info;

        root->right = deleteNode(root->right, temp->info);
    }
    return root;
}

int main()
{
    int choice, x;
    while (1)
    {
        system("CLS");

        printf("Inorder\n");
        inorder(root);
        printf("\nPreorder\n");
        preorder(root);
        printf("\nPostorder\n");
        postorder(root);

        printf("\nBinary Search Tree\n");
        printf("Press 1 to insert\n");
        printf("Press 2 to delete\n");
        printf("Press 0 to exit\n");
        printf("Choice? ");
        scanf("%d", &choice);
        switch (choice)
        {
        case 1:
            printf("Element? ");
            scanf("%d", &x);
            root = insertNode(root, x);
            break;
        case 2:
            printf("Element? ");
            scanf("%d", &x);
            root = deleteNode(root, x);
            break;
        case 0:
            exit(0);
        default:
            printf("Wrong Choice");
        }
    }
    return 0;
}

File: data/Algorithms/Binary Search Tree.py
class Node():
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None
        
    def __str__(self):
        return f"{self.data}"
        
def find(root, data):
    currentNode = root
    while currentNode is not None and data != currentNode.data:
        if data> currentNode.data:
            currentNode = currentNode.right
        else:
            currentNode = currentNode.left
    return currentNode

def findMin(root):
    currentNode = root
    if currentNode.left is None:
        return currentNode
    else:
        return findMin(currentNode.left)
    
def findMax(root):
    currentNode = root
    if currentNode.right is None:
        return currentNode
    else:
        return findMax(currentNode.right)
    
def successorBST(root):
    currentNode = None
    if root.right is not None:
        currentNode = root.right
        while currentNode.left is not None:
            currentNode = currentNode.left
    return currentNode

def predecessorBST(root):
    currentNode = None
    if root.left is not None:
        currentNode = root.left
        while currentNode.right is not None:
            currentNode = currentNode.right
    return currentNode

def insert(node, key): 
    if node is None: 
        return Node(key) 
    if key < node.data: 
        node.left = insert(node.left, key) 
    else: 
        node.right = insert(node.right, key) 
    return node 

def inorderTraversal(node):
    res = []
    if node:
        res = inorderTraversal(node.left)
        res.append(node.data)
        res = res + inorderTraversal(node.right)
    return res

def preorderTraversal(node):
    res = []
    if node:
        res.append(node.data)
        res = res + preorderTraversal(node.left)
        res = res + preorderTraversal(node.right)
    return res

def postorderTraversal(node):
    res = []
    if node:
        res = postorderTraversal(node.left)
        res = res + postorderTraversal(node.right)
        res.append(node.data)
    return res

def deleteNode(root, key): 
    if root is None: 
        return root   
    if key < root.data: 
        root.left = deleteNode(root.left, key)  
    elif(key > root.data): 
        root.right = deleteNode(root.right, key) 
    else: 
        if root.left is None : 
            temp = root.right  
            root = None 
            return temp    
        elif root.right is None : 
            temp = root.left  
            root = None
            return temp 
        temp = findMin(root.right)   
        root.data = temp.data  
        root.right = deleteNode(root.right , temp.data) 
    return root  

if __name__ == "__main__":
    root = Node(10)
    root = insert(root, 50) 
    root = insert(root, 30) 
    root = insert(root, 20) 
    root = insert(root, 40) 
    root = insert(root, 70) 
    root = insert(root, 60) 
    root = insert(root, 80)
    print(inorderTraversal(root))
    root = deleteNode(root, 20) 
    print(inorderTraversal(root))